### PR TITLE
Server timeout and Keep Alive interval config

### LINKED ISSUE #29296: Server timeout and Keep Alive interval
* https://github.com/dotnet/aspnetcore/issues/44742
* https://github.com/dotnet/aspnetcore/pull/46065

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: 0969556b-b77e-bdb8-8e20-fd6971be0c25
* Version Independent ID: d17e4c28-ea82-0dd8-295d-19f6ada147ea
* Content: [ASP.NET Core Blazor SignalR guidance](https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/signalr?view=aspnetcore-7.0)
* Content Source: [aspnetcore/blazor/fundamentals/signalr.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/blazor/fundamentals/signalr.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-fundamentals**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #29296

Surayya ...

* I searched through the SignalR node looking for update opportunities, but it doesn't look like explicit config for the server timeout or Keep-Alive interval with the JS client or .NET client are actually shown, the settings only seem to be named. I've tried to show the JS client code in the *Release notes* article correctly, but I won't be surprised if it's incorrect. I'm not as familiar with the general SignalR scenario compared to Blazor.
* Ignore the little bit of version flipping going on where the Blazor content flips in and out of 8.0 to cover this. These pair of *Blazor* > *Host and deploy* node topics haven't been addressed yet for versioning. I'll fix up the versioning later.

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/blazor/fundamentals/signalr.md](https://github.com/dotnet/AspNetCore.Docs/blob/bbbd9de4091ce197b66f8b3cab12267a1c381e57/aspnetcore/blazor/fundamentals/signalr.md) | [ASP.NET Core Blazor SignalR guidance](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/signalr?branch=pr-en-us-29297) |
| [aspnetcore/blazor/host-and-deploy/server.md](https://github.com/dotnet/AspNetCore.Docs/blob/bbbd9de4091ce197b66f8b3cab12267a1c381e57/aspnetcore/blazor/host-and-deploy/server.md) | [Host and deploy ASP.NET Core Blazor Server](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/host-and-deploy/server?branch=pr-en-us-29297) |
| [aspnetcore/blazor/host-and-deploy/webassembly.md](https://github.com/dotnet/AspNetCore.Docs/blob/bbbd9de4091ce197b66f8b3cab12267a1c381e57/aspnetcore/blazor/host-and-deploy/webassembly.md) | [Host and deploy ASP.NET Core Blazor WebAssembly](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/host-and-deploy/webassembly?branch=pr-en-us-29297) |
| [aspnetcore/release-notes/aspnetcore-8.0.md](https://github.com/dotnet/AspNetCore.Docs/blob/bbbd9de4091ce197b66f8b3cab12267a1c381e57/aspnetcore/release-notes/aspnetcore-8.0.md) | [aspnetcore/release-notes/aspnetcore-8.0](https://review.learn.microsoft.com/en-us/aspnet/core/release-notes/aspnetcore-8.0?branch=pr-en-us-29297) |

### CODE REVIEW

**Review by JzumHub — APPROVED** [review]

**Comment by guardrex** [comment]
Ping @surayya-MS ... Just making sure that you saw this ... no rush.

File: aspnetcore/blazor/fundamentals/signalr.md [hunk 1]
```diff
@@ -443,7 +443,25 @@ Configure the following values for the client:
 * `serverTimeoutInMilliseconds`: The server timeout in milliseconds. If this timeout elapses without receiving any messages from the server, the connection is terminated with an error. The default timeout value is 30 seconds. The server timeout should be at least double the value assigned to the Keep-Alive interval (`keepAliveIntervalInMilliseconds`).
 * `keepAliveIntervalInMilliseconds`: Default interval at which to ping the server. This setting allows the server to detect hard disconnects, such as when a client unplugs their computer from the network. The ping occurs at most as often as the server pings. If the server pings every five seconds, assigning a value lower than `5000` (5 seconds) pings every five seconds. The default value is 15 seconds. The Keep-Alive interval should be less than or equal to half the value assigned to the server timeout (`serverTimeoutInMilliseconds`).
```
Comment (surayya-MS) [inline aspnetcore/blazor/fundamentals/signalr.md:444]: Let's rewrite this section according to the current design. For TypeScript client new design is
```diff
export class HubConnectionBuilder
{
+    public withServerTimeout(milliseconds: number): HubConnectionBuilder;
+    public withKeepAliveInterval(milliseconds: number): HubConnectionBuilder;
}
```
Let's rephrase these portions of text where `serverTimeoutInMilliseconds` and `keepAliveIntervalInMilliseconds` are described. Replace with description for `withServerTimeout` and `withKeepAliveInterval` methods and keep the description of what  `serverTimeoutInMilliseconds` and `keepAliveIntervalInMilliseconds` actually are. Also let's add
`withServerTimeout` and `withKeepAliveInterval` methods configure `HubConnectionBuilder` and return the same instance of `HubConnectionBuilder` so they can be chained.

File: aspnetcore/blazor/fundamentals/signalr.md [hunk 2]
```diff
@@ -443,7 +443,25 @@ Configure the following values for the client:
...
+
+```html
+<script src="_framework/blazor.{HOSTING MODEL}.js" autostart="false"></script>
+<script>
+  Blazor.start({
+    configureSignalR: function (builder) {
+      builder.serverTimeoutInMilliseconds = 30000;
+      builder.keepAliveIntervalInMilliseconds = 15000;
+    }
+  });
```
Comment (surayya-MS) [inline aspnetcore/blazor/fundamentals/signalr.md:458]: It should be 
```ts
Blazor.start({
    configureSignalR: function (builder) {
        builder.withServerTimeout(30000).withKeepAliveInterval(15000);
    };
});
```

File: aspnetcore/blazor/fundamentals/signalr.md [hunk 3]
```diff
@@ -443,7 +443,25 @@ Configure the following values for the client:
...
+</script>
+```
+
+:::moniker-end
+
+:::moniker range="< aspnetcore-8.0"
 
 ```html
 <script src="_framework/blazor.{HOSTING MODEL}.js" autostart="false"></script>
```
Comment (surayya-MS) [inline aspnetcore/blazor/fundamentals/signalr.md:467]: Let's remove this section with a workaround

File: aspnetcore/blazor/fundamentals/signalr.md [hunk 4]
```diff
@@ -461,9 +479,36 @@ The following example for the `Pages/_Host.cshtml` file (Blazor Server, all vers
...
+:::moniker range=">= aspnetcore-8.0"
+
+When creating a hub connection in a component, set the <xref:Microsoft.AspNetCore.SignalR.Client.HubConnection.ServerTimeout> (default: 30 seconds) and <xref:Microsoft.AspNetCore.SignalR.Client.HubConnection.KeepAliveInterval> (default: 15 seconds) on the <xref:Microsoft.AspNetCore.SignalR.Client.HubConnectionBuilder>. Set the <xref:Microsoft.AspNetCore.SignalR.Client.HubConnection.HandshakeTimeout> (default: 15 seconds) on the built <xref:Microsoft.AspNetCore.SignalR.Client.HubConnection>. The following example, based on the `Index` component in the [SignalR with Blazor tutorial](xref:blazor/tutorials/signalr-blazor), shows the assignment of default values:
```
Comment (surayya-MS) [inline aspnetcore/blazor/fundamentals/signalr.md:488]: @javiercn , was it this [chat tutorial](https://learn.microsoft.com/en-us/aspnet/core/blazor/tutorials/signalr-blazor?view=aspnetcore-6.0&tabs=visual-studio&pivots=server) that we wanted to delete from docs?

File: aspnetcore/release-notes/aspnetcore-8.0.md [hunk 5]
```diff
@@ -28,6 +28,100 @@ This article is under development and not complete. More information may be foun
...
+
+The following example shows the ***new approach*** for assigning values that are double the default values in ASP.NET Core 8.0 or later:
+
+```javascript
+Blazor.start({
+  configureSignalR: function (builder) {
+    builder.serverTimeoutInMilliseconds = 60000;
+    builder.keepAliveIntervalInMilliseconds = 30000;
+  }
+});
```
Comment (surayya-MS) [inline aspnetcore/release-notes/aspnetcore-8.0.md:89]:
```suggestion
Blazor.start({
  configureSignalR: function (builder) {
    builder.withServerTimeout(60000).withKeepAliveInterval(30000);
  }
});
```

**Comment by guardrex** [comment]
@surayya-MS ... I was working through the changes, and I need to double-check this one ...

> Let's remove the following section. This was a workaround to set serverTimeoutInMilliseconds and keepAliveIntervalInMilliseconds 

... because if we do that it isn't just this section. The workaround is also part of the *Global deployment and connection failures* section, and it's going to get chopped for that coverage prior to 8.0, too. We'll be telling devs for <8.0 that they need to change it, but they can't because we don't support the workaround.

**Comment by surayya-MS** [comment]
> ... because if we do that it isn't just this section. The workaround is also part of the Global deployment and connection failures section, and it's going to get chopped for that coverage prior to 8.0, too. We'll be telling devs for <8.0 that they need to change it, but they can't because we don't support the workaround.

@guardrex sorry I previously missed the <8.0 part. I assumed that this is for aspnetcore-8.0. You are right we shouldn't remove it then.


>Also note that I don't follow the opening paragraph instructions ("rephrase"/"replace" but "keep") ... Do you mean that in the text only the names should change for 8.0 or later? ... and "actually are" implies that the current text is incorrect, but I think this was all reviewed by Javier (in ancient Blazor history 👴😆).

I meant let's replace descriptions for `serverTimeoutInMilliseconds` and `keepAliveIntervalInMilliseconds` with descriptions for `withServerTimeout` and `withKeepAliveInterval`. And in there describe what `serverTimeoutInMilliseconds` and `keepAliveIntervalInMilliseconds` actually are. No, it wasn't implied that `serverTimeoutInMilliseconds` and `keepAliveIntervalInMilliseconds` are wrong. I just meant it would be better to describe what `withServerTimeout` and `withKeepAliveInterval` are setting exactly.

I'll write a follow up comment for that

File: aspnetcore/blazor/fundamentals/signalr.md (hunk 1 above)
Comment (surayya-MS) [reply aspnetcore/blazor/fundamentals/signalr.md:444]: Ignore previous comment

**Comment by guardrex** [comment]
Note that the last commit is just a tiny NIT ... I'm merely hyphenating "Keep-Alive" for consistency. That's all that's on the commit.

**Comment by guardrex** [comment]
I didn't change the first part tho ... the paragraph. Do you want to supply the text, or do you want me to try and fix that and ping u back?

**Comment by guardrex** [comment]
Oh! I see ........

> Update: No follow up comments. Looks good to me!

Gotcha! Ok, good to merge then! 🎉

**Comment by guardrex** [comment]
... but I'll circle around later if you/Javier intend to pull the Blazor-SignalR tutorial from the doc set per https://github.com/dotnet/AspNetCore.Docs/pull/29297#discussion_r1214356841.

**Comment by surayya-MS** [comment]
> I didn't change the first part tho ... the paragraph. Do you want to supply the text, or do you want me to try and fix that and ping u back?

No need to change the paragraph because of the workaround example for > 8.0. The descriptions for `serverTimeoutInMilliseconds` and `keepAliveIntervalInMilliseconds` are needed there.

**Comment by guardrex** [comment]
> No need to change the paragraph because of the workaround example for > 8.0. The descriptions for serverTimeoutInMilliseconds and keepAliveIntervalInMilliseconds are needed there.

Yes ... but we might want to organize this a bit differently. Perhaps, I should move that API down into the < 8.0 block and have a different version of it that only explains `withServerTimeout` and `withKeepAliveInterval` in the 8.0 or later block.

Let me put that on a commit and you can see how it composes that way. *Stand-by ..................................*

**Comment by surayya-MS** [comment]
LGTM!

**Pull request merged**